lib/ansible/modules/cloud/google/gcp_dns_managed_zone.py

```diff
@@ -0,0 +1,238 @@
+from ansible.module_utils.gcp_utils import navigate_hash, GcpAuthentication, GcpModule
+import json
+
+################################################################################
+# Main
+################################################################################
+
+
+def main():
+    """Main function"""
+
+    module = GcpModule(
+        argument_spec=dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            description=dict(type='str'),
+            dns_name=dict(type='str'),
+            name=dict(required=True, type='str'),
+            name_server_set=dict(type='list'),
+        )
+    )
+
+    state = module.params['state']
+    kind = 'dns#managedZone'
+
+    fetch = fetch_resource(module, self_link(module), kind)
+    changed = False
+
+    if fetch:
+        if state == 'present':
+            if is_different(module, fetch):
+                fetch = update(module, self_link(module), kind)
+        else:
+            delete(module, self_link(module), kind)
+            fetch = {}
+            changed = True
+    else:
+        if state == 'present':
+            fetch = create(module, collection(module), kind)
+            changed = True
+
+    if fetch:
+        fetch.update({'changed': changed})
+    else:
+        fetch = {'changed': changed}
+
+    module.exit_json(**fetch)
+
+
+def create(module, link, kind):
+    auth = GcpAuthentication(module)
+    return return_if_object(module, auth.session().post(link, json=resource_to_request(module)), kind)
+
+
+def update(module, link, kind):
+    module.fail_json(msg="ManagedZone cannot be edited")
+
+
+def delete(module, link, kind):
+    auth = GcpAuthentication(module)
+    return return_if_object(module, auth.session().delete(link), kind)
+
+
+def resource_to_request(module):
+    request = {
+        u'kind': 'dns#managedZone',
+        u'description': module.params['description'],
+        u'dnsName': module.params['dns_name'],
+        u'name': module.params['name'],
+        u'nameServerSet': module.params['name_server_set'],
+    }
+    return_vals = {}
+    for k, v in request.items():
+        if v:
+            return_vals[k] = v
+
+    return return_vals
+
+
+def fetch_resource(module, link, kind):
+    auth = GcpAuthentication(module)
+    return return_if_object(module, auth.session().get(link), kind)
+
+
+def self_link(module):
+    return "https://www.googleapis.com/dns/v1/projects/{project}/managedZones/{name}".format(**module.params)
+
+
+def collection(module):
+    return "https://www.googleapis.com/dns/v1/projects/{project}/managedZones".format(**module.params)
+
+
+def return_if_object(module, response, kind):
+    # If not found, return nothing.
+    if response.status_code == 404:
+        return None
+
+    # If no content, return nothing.
+    if response.status_code == 204:
+        return None
+
+    try:
+        response.raise_for_status
+        result = response.json()
+    except Exception as inst:
```

Comment:
I'm not big on returning `Exception` here I'd rather be specific to JSON-ish exceptions, like:

```
try:
    pass # blah blah code
# on Python 3+, json.decoder.JSONDecodeError is raised for bad
# JSON. On 2.x it's a ValueError
except getattr(json.decoder, 'JSONDecodeError', ValueError) as e:
    module.fail_json(...)
```

Reply:
And then for requests-status issues, have a separate clause.